keps/sig-node/3702-ima-namespace-support/README.md

```diff
@@ -0,0 +1,263 @@
+
+IMA and EVM can use a TPM chip as a hardware root of trust. Hence we can verify images against a set of golden hash values, as well as avoiding any further changes to the overlayfs to intercept calls and check the integrity of files. -->
+
+### User Stories (Optional)
+
+#### Story 1
+As a cluster admin, I want to detect undesired file changes, so that I can take out pods that have been compromised.
+
+An intruder perform some malicious changes inside a certain pod's files. The system should be able to detect those changed and alert about the inconsistent pod. The remote attestation framework could keep this pod running or make a copy (for forensic analysis) and delete it.
+
+#### Story 2
+As a cluster admin, I want to deploy a only proven and certified pods, so that I can comply with internal policies as well as security regulations.
+
+Let's say that we are working in a high security environment where only approved images can be deployed. In this scenario we can make sure that the pod deploy used an imaged that hasn't been tampered.
+
+#### Story 3
+As a cluster admin, I want to deny access to certain files inside the pod, so that a potential intruder can't access sensitive information.
+
+In some cases, we should not even allow root to modify certain files inside the container.
+
+### Notes/Constraints/Caveats (Optional)
+We need to enable IMA in the kernel and container runtimes (runC, CRI-O, docker, containerd, etc.). 
+
+### Risks and Mitigations
+
+  
+
+## Design Details
+  
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+In order for the this feature to work, the nodes where the pods will be deploy should have IMA enabled and a recent kernel that supports IMA namespaces (WIP, it should be merged soon). The nodes should also have a TPM chip. We could use nodAffinity of labels and annotations in nodes in order to select where to deploy the pods.
+
+The linux kernel IMA namespace support is based on user namespaces. Therefore, the container runtime should first create a user namespace and then create an IMA namespaces. In order to use IMA namespaces it is necessary to enable user namespaces as well.
+
+Should we enable IMA namespaces by default when enabling user namespaces?
+
+
+There will be a CRI API change which will allow the pod to use IMA namespaces and specify the namespace policy.
+
+  
+
+### Linux kernel
+
+  
+
+IMA is only available in Linux hosts and Linux containers. Unfortunately, IMA is not a separate namespace, which is needed in order to isolate it and be used inside containers. Upcoming kernel patches should add support for IMA namespaces.
```

Comment:
This seems confusing, you say it is not a separate namespace however the KEP talks all the time about the IMA namespace.

I guess you refer to the fact that won't be unshared? Not sure that is relevant here, but if you need to go into those details, please don't say it is not a namespace here and that it is a namespace in the rest of the KEP

Reply:
Currently IMA namespace cannot "live" without any user namespace. IOW one user_ns -- one IMA namespace. Is there any initial user namespace in Kuber by defaut? @asierHuawei @rata

Reply:
Yes, kubernetes 1.25 supports userns for stateless pods (this is alpha). I'll rework this to depend on idmap mounts for stateless pods soon (will need Linux >=6.3) and then add support for stateful pods

Reply:
@rata Hm... sounds good, good news.

Reply:
@DenisSemakin cool. Do you need more patches to the kernel besides the ima ns patchset?

Reply:
@rata If we're talking about IMA namespaces as a standalone feature then the answer is... I guess it's "No". All patches provided by Stefan cover the basic functionality. But there's still opened question I see here: "will it be enough those patches for our purpose?"... 
In case of remote attestation and pod (container) verification the answer is definitely "Yes". Here is a link to which patches we would like to add for this: https://lore.kernel.org/linux-integrity/20221031025507.298323-1-denis.semakin@huawei.com/